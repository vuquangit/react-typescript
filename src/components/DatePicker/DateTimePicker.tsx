import React, { FC, MutableRefObject, useRef, useState } from 'react'
import Button from '../Button'

import {
  DateTimePickerWrapper,
  DTPDatetimeField,
  DTPDatetimeSlash,
  DTPFormControl,
  DTPInput,
  DTPLabel,
} from './DateTimePicker.styled'
import { BaseDateTimePickerProps } from './DateTimePicker.types'

const FIELD_DAY = 'day'
const FIELD_MONTH = 'month'
const FIELD_YEAR = 'year'
type IDatetimeField = typeof FIELD_DAY | typeof FIELD_MONTH | typeof FIELD_YEAR

const DateTimePicker: FC<BaseDateTimePickerProps> = ({ type = 'date' }) => {
  const dayInput = useRef(document.createElement('input'))
  const monthInput = useRef(document.createElement('input'))
  const yearInput = useRef(document.createElement('input'))

  const addLeadingZeros = (
    val: number,
    field: IDatetimeField,
    isMonthInit = false
  ): string => {
    let minLength = 0
    if (field === FIELD_DAY) minLength = 2
    else if (field === FIELD_MONTH) minLength = 2
    else minLength = 4

    const val2Str = (isMonthInit ? 1 : 0) + val + ''
    if (val2Str.length < minLength)
      return new Array(minLength - val2Str.length).fill(0).join('') + val2Str

    return val2Str
  }

  const [datetime, setDatetime] = useState(() => {
    const today = new Date()

    return {
      day: addLeadingZeros(today.getDate(), FIELD_DAY),
      month: addLeadingZeros(today.getMonth(), FIELD_MONTH, true),
      year: addLeadingZeros(today.getUTCFullYear(), FIELD_YEAR),
    }
  })

  const onDatetimeFieldClick = (
    fieldRef: MutableRefObject<HTMLInputElement>
  ) => {
    if (fieldRef && fieldRef.current) fieldRef.current.select()
  }

  const changeValueCircle = (val: number, maxVal: number) => {
    const newVal = ((val === 0 ? val - 1 : val) + maxVal + 1) % (maxVal + 1)

    return newVal === 0 ? newVal + 1 : newVal
  }

  const changeDatetime = (
    e: any,
    field: IDatetimeField,
    val: string
  ): string => {
    let str2Num = parseInt(val)

    if (e.keyCode === 38) {
      if (field === FIELD_DAY) str2Num = changeValueCircle(str2Num + 1, 31)
      else if (field === FIELD_MONTH)
        str2Num = changeValueCircle(str2Num + 1, 12)
      else if (field === FIELD_YEAR) str2Num += 1
    } else if (e.keyCode === 40) {
      if (field === FIELD_DAY) str2Num = changeValueCircle(str2Num - 1, 31)
      else if (field === FIELD_MONTH)
        str2Num = changeValueCircle(str2Num - 1, 12)
      else if (field === FIELD_YEAR && str2Num > 1) str2Num -= 1
    }

    return addLeadingZeros(str2Num, field)
  }

  const onDatetimeFieldKeydown = (e: any, field: IDatetimeField) => {
    // https://reactjs.org/docs/legacy-event-pooling.html
    // Prevents React from resetting its properties:
    e.persist()

    if (field === FIELD_DAY && dayInput) {
      setDatetime((prev) => ({
        ...prev,
        day: changeDatetime(e, field, dayInput.current.value),
      }))
    } else if (field === FIELD_MONTH && monthInput) {
      setDatetime((prev) => ({
        ...prev,
        month: changeDatetime(e, field, monthInput.current.value),
      }))
    } else if (field === FIELD_YEAR && yearInput) {
      setDatetime((prev) => ({
        ...prev,
        year: changeDatetime(e, field, yearInput.current.value),
      }))
    }
  }

  // input change
  const isNumber = (val: any): boolean => val.match('/[0-9]/')

  const onDatetimeFieldChange = (e: any, field: IDatetimeField) => {
    console.log(field, e.target.value)

    const val = e.target.value
    if (val !== null && !isNumber(val)) return

    const valInt = parseInt(val)

    if (field === 'day') {
      if (0 <= valInt && valInt < 9) {
        setDatetime((prev) => ({ ...prev, day: addLeadingZeros(val, field) }))
      } else {
        setDatetime((prev) => ({
          ...prev,
          day: prev.day + addLeadingZeros(val, field),
        }))
      }
    }

    // onDatetimeFieldClick(dayInput)
  }

  return (
    <DateTimePickerWrapper>
      <DTPLabel htmlFor="date">Datetime picker:</DTPLabel>
      <DTPFormControl>
        <DTPInput>
          <DTPDatetimeField
            type="text"
            // readOnly
            value={datetime.day}
            ref={dayInput}
            onClick={() => onDatetimeFieldClick(dayInput)}
            onKeyDown={(e) => onDatetimeFieldKeydown(e, FIELD_DAY)}
            onChange={(e) => onDatetimeFieldChange(e, FIELD_DAY)}
          />
          <DTPDatetimeSlash>/</DTPDatetimeSlash>
          <DTPDatetimeField
            type="text"
            // readOnly
            value={datetime.month}
            ref={monthInput}
            onClick={() => onDatetimeFieldClick(monthInput)}
            onKeyDown={(e) => onDatetimeFieldKeydown(e, FIELD_MONTH)}
            onChange={(e) => onDatetimeFieldChange(e, FIELD_MONTH)}
          />
          <DTPDatetimeSlash>/</DTPDatetimeSlash>
          <DTPDatetimeField
            type="text"
            // readOnly
            ref={yearInput}
            value={datetime.year}
            onClick={() => onDatetimeFieldClick(yearInput)}
            onKeyDown={(e) => onDatetimeFieldKeydown(e, FIELD_YEAR)}
            onChange={(e) => onDatetimeFieldChange(e, FIELD_YEAR)}
          />
        </DTPInput>
        <Button>Calendar icon</Button>
      </DTPFormControl>
    </DateTimePickerWrapper>
  )
}

export default DateTimePicker
