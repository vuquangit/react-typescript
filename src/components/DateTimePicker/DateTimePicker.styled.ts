import styled from 'styled-components'
import { space } from 'styled-system'

export const DateTimePickerWrapper = styled('div')`
  ${space};
  // border: 0;
  // margin: 0;
  // display: inline-flex;
  // padding: 0;
  // position: relative;
  // min-width: 0;
  // flex-direction: column;
  // vertical-align: top;
`
export const DTPLabel = styled('label')`
  // display: block;
  // top: 0;
  // left: 0;
  // position: absolute;
  // transition: color 200ms cubic-bezier(0, 0, 0.2, 1) 0ms,
  //   transform 200ms cubic-bezier(0, 0, 0.2, 1) 0ms;
  // transform: translate(0, 1.5px) scale(0.75);
  // transform-origin: top left;
`
export const DTPFormControl = styled('div')`
  display: flex;
  align-items: center;
  position: relative;
`

export const DTPInput = styled('div')`
  display: flex;
  // border: 1px #000 solid;

  &:after {
    left: 0;
    right: 0;
    bottom: 0;
    content: '';
    position: absolute;
    // transform: scaleX(0);
    transition: transform 200ms cubic-bezier(0, 0, 0.2, 1) 0ms;
    border-bottom: 2px solid #90caf9;
    pointer-events: none;
  }
`

export const DTPDatetimeField = styled('input')`
  // display: inline;
  -webkit-user-modify: read-only !important;
  padding: 0;
  margin: 0;
  border: 0;
  // user-select: all;
  cursor: default;

  &::-webkit-inner-spin-button,
  &::-webkit-outer-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }

  &:focus {
    background-color: rgb(153, 200, 255);
    color: rgb(0, 0, 0);
    outline: none;
  }
`

export const DTPDatetimeSlash = styled('div')`
  cursor: default;
`
