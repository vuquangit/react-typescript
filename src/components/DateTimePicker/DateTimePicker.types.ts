import { SpaceProps } from 'styled-system'

export interface IMDateTimePickerProps {
  type: 'date' | 'datetime-local'
}
export declare type BaseDateTimePickerProps = React.ButtonHTMLAttributes<
  HTMLInputElement
> &
  IMDateTimePickerProps &
  SpaceProps

export declare type DatetimeField = React.ButtonHTMLAttributes<HTMLInputElement>
