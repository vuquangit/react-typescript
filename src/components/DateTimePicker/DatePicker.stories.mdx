import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks'
import { action } from '@storybook/addon-actions'

import AppWrapper from '@/test/supports/AppWrapper'
import DateTimePicker from '.'
import Space from 'stories/components/styled-system/Space.mdx'
import Styles from 'stories/styles.mdx'

<Meta title="Story/Components/DateTimePicker" component={DateTimePicker} />

<Styles />

# DateTimePicker

With `MDX` we can define a story for `DateTimePicker` right in the middle of our
Markdown documentation.

export const TemplateDate = (args) => <AppWrapper><DateTimePicker {...args} /></AppWrapper>

<Canvas>
  <Story
    name="Date Type"
    args={{ typePicker: 'date' }}
  >
    {TemplateDate.bind({})}
  </Story>
   <Story
    name="Date-time Local Type"
    args={{ typePicker: 'datetime-local' }}
  >
    {TemplateDate.bind({})}
  </Story>
    <Story
    name="Time Type"
    args={{ typePicker: 'time' }}
  >
    {TemplateDate.bind({})}
  </Story>
</Canvas>

<div class="subheading">Props</div>

<Space />
