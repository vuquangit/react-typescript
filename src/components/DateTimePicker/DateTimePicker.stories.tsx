import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'

import AppWrapper from '@/test/supports/AppWrapper'
import DateTimePicker from '.'
import { BaseDateTimePickerProps } from './DateTimePicker.types'

export default {
  title: 'Story/Components/DateTimePicker',
  component: DateTimePicker,
  argTypes: {},
} as Meta

const TemplateDateTimePicker: Story<BaseDateTimePickerProps> = (args) => (
  <AppWrapper>
    <DateTimePicker {...args}>Component story</DateTimePicker>
  </AppWrapper>
)

export const DateType = TemplateDateTimePicker.bind({})
DateType.args = {
  typePicker: 'date',
}

export const DateTimeLocalType = TemplateDateTimePicker.bind({})
DateTimeLocalType.args = {
  typePicker: 'datetime-local',
}

export const TimeType = TemplateDateTimePicker.bind({})
TimeType.args = {
  typePicker: 'time',
}
