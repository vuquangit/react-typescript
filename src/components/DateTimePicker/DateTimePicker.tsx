import React, { FC, MutableRefObject, useEffect, useRef, useState } from 'react'
import Button from '../Button'

import {
  DateTimePickerWrapper,
  DTPDatetimeField,
  DTPDatetimeSlash,
  DTPFormControl,
  DTPInput,
  DTPLabel,
} from './DateTimePicker.styled'
import { BaseDateTimePickerProps } from './DateTimePicker.types'

const FIELD_DAY = 'day'
const FIELD_MONTH = 'month'
const FIELD_YEAR = 'year'
type IDatetimeField = typeof FIELD_DAY | typeof FIELD_MONTH | typeof FIELD_YEAR

const DateTimePicker: FC<BaseDateTimePickerProps> = ({ type }) => {
  const dayInput = useRef(document.createElement('input'))
  const monthInput = useRef(document.createElement('input'))
  const yearInput = useRef(document.createElement('input'))

  const addLeadingZeros = (
    val: number,
    field: IDatetimeField,
    isMonthInit = false
  ): string => {
    let minLength = 0
    if (field === FIELD_DAY) minLength = 2
    else if (field === FIELD_MONTH) minLength = 2
    else minLength = 4

    const val2Str = (isMonthInit ? 1 : 0) + val + ''
    if (val2Str.length < minLength)
      return new Array(minLength - val2Str.length).fill(0).join('') + val2Str

    return val2Str
  }

  const [datetime, setDatetime] = useState(() => {
    const today = new Date()

    return {
      day: addLeadingZeros(today.getDate(), FIELD_DAY),
      month: addLeadingZeros(today.getMonth(), FIELD_MONTH, true),
      year: addLeadingZeros(today.getUTCFullYear(), FIELD_YEAR),
    }
  })

  const onDatetimeFieldClick = (
    fieldRef: MutableRefObject<HTMLInputElement>
  ) => {
    if (fieldRef && fieldRef.current) fieldRef.current.select()
    resetInputCount()
  }

  const changeValueCircle = (val: number, maxVal: number) => {
    const newVal = ((val === 0 ? val - 1 : val) + maxVal + 1) % (maxVal + 1)

    return newVal === 0 ? newVal + 1 : newVal
  }

  const changeDatetime = (
    e: any,
    field: IDatetimeField,
    val: string
  ): string => {
    let str2Num = parseInt(val)

    if (e.keyCode === 38) {
      if (field === FIELD_DAY) str2Num = changeValueCircle(str2Num + 1, 31)
      else if (field === FIELD_MONTH)
        str2Num = changeValueCircle(str2Num + 1, 12)
      else if (field === FIELD_YEAR) str2Num += 1
    } else if (e.keyCode === 40) {
      if (field === FIELD_DAY) str2Num = changeValueCircle(str2Num - 1, 31)
      else if (field === FIELD_MONTH)
        str2Num = changeValueCircle(str2Num - 1, 12)
      else if (field === FIELD_YEAR && str2Num > 1) str2Num -= 1
    }

    return addLeadingZeros(str2Num, field)
  }

  const onDatetimeFieldKeydown = (
    e: React.KeyboardEvent<HTMLInputElement>,
    field: IDatetimeField
  ) => {
    // https://reactjs.org/docs/legacy-event-pooling.html
    // Prevents React from resetting its properties:
    e.persist()

    if (field === FIELD_DAY && dayInput) {
      setDatetime((prev) => ({
        ...prev,
        day: changeDatetime(e, field, dayInput.current.value),
      }))
    } else if (field === FIELD_MONTH && monthInput) {
      setDatetime((prev) => ({
        ...prev,
        month: changeDatetime(e, field, monthInput.current.value),
      }))
    } else if (field === FIELD_YEAR && yearInput) {
      setDatetime((prev) => ({
        ...prev,
        year: changeDatetime(e, field, yearInput.current.value),
      }))
    }
  }

  // input change
  const isNumber = (val: any): boolean => val.match(/[0-9]/)

  const [dayInputCount, setDayInputCount] = useState(0)
  const [monthInputCount, setMonthInputCount] = useState(0)
  const [yearInputCount, setYearInputCount] = useState(0)

  const onDatetimeFieldChange = (
    e: React.ChangeEvent<HTMLInputElement>,
    field: IDatetimeField
  ) => {
    const val: string = e.target.value.slice(-1)

    if (val === null || !isNumber(val)) return

    const valInt = parseInt(val)

    if (field === FIELD_DAY) {
      if (dayInputCount === 0 && valInt <= 3) {
        setDatetime((prev) => ({
          ...prev,
          day: addLeadingZeros(valInt, field),
        }))
        setDayInputCount(1)
      } else {
        setDatetime((prev) => ({
          ...prev,
          day:
            dayInputCount === 1
              ? prev.day.slice(-1) + valInt
              : addLeadingZeros(valInt, field),
        }))

        setDayInputCount(0)
        monthInput.current.focus()
      }
    } else if (field === FIELD_MONTH) {
      if (monthInputCount === 0) {
        setDatetime((prev) => ({
          ...prev,
          month: addLeadingZeros(valInt, field),
        }))

        if (valInt > 2) {
          setMonthInputCount(0)
          yearInput.current.focus()
        } else {
          setMonthInputCount(1)
        }
      } else if (monthInputCount > 0 && valInt <= 2) {
        setDatetime((prev) => ({
          ...prev,
          month: prev.month.slice(-1) + valInt,
        }))

        setMonthInputCount(0)
        yearInput.current.focus()
      }
    } else if (field === FIELD_YEAR) {
      const nextYear =
        (yearInputCount < 6
          ? datetime.year.slice(-(yearInputCount + 1))
          : datetime.year.slice(-5)) + valInt

      setDatetime((prev) => ({
        ...prev,
        year: addLeadingZeros(parseInt(nextYear), field),
      }))

      setYearInputCount(yearInputCount + 1)
    }
  }

  const resetInputCount = () => {
    setDayInputCount(0)
    setMonthInputCount(0)
    setYearInputCount(0)
  }

  return (
    <DateTimePickerWrapper>
      <DTPLabel htmlFor="date">Datetime picker:</DTPLabel>
      <DTPFormControl>
        <DTPInput>
          <DTPDatetimeField
            type="text"
            // readOnly
            value={datetime.day}
            ref={dayInput}
            onClick={() => onDatetimeFieldClick(dayInput)}
            onKeyDown={(e) => onDatetimeFieldKeydown(e, FIELD_DAY)}
            onChange={(e) => onDatetimeFieldChange(e, FIELD_DAY)}
          />
          <DTPDatetimeSlash>/</DTPDatetimeSlash>
          <DTPDatetimeField
            type="text"
            // readOnly
            value={datetime.month}
            ref={monthInput}
            onClick={() => onDatetimeFieldClick(monthInput)}
            onKeyDown={(e) => onDatetimeFieldKeydown(e, FIELD_MONTH)}
            onChange={(e) => onDatetimeFieldChange(e, FIELD_MONTH)}
          />
          <DTPDatetimeSlash>/</DTPDatetimeSlash>
          <DTPDatetimeField
            type="text"
            // readOnly
            ref={yearInput}
            value={datetime.year}
            onClick={() => onDatetimeFieldClick(yearInput)}
            onKeyDown={(e) => onDatetimeFieldKeydown(e, FIELD_YEAR)}
            onChange={(e) => onDatetimeFieldChange(e, FIELD_YEAR)}
          />
        </DTPInput>
        <Button>Calendar icon</Button>
      </DTPFormControl>
    </DateTimePickerWrapper>
  )
}

export default DateTimePicker
