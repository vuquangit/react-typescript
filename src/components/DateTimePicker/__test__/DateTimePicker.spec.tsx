/**
 *
 * Tests for DateTimePicker
 *
 */

import React, { useReducer } from 'react'
import { render /* fireEvent */ } from '@testing-library/react'

import DateTimePicker from '..'
import AppWrapper from '@/test/supports/AppWrapper'

describe('App SwitchTheme', () => {
  it('Test SwitchTheme with default', async () => {
    const { container, getByTestId } = render(
      <AppWrapper>
        <DateTimePicker />
      </AppWrapper>
    )

    expect(container).toMatchSnapshot()
  })
})
