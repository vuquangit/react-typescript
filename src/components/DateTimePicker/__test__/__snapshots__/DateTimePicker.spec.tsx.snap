// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App SwitchTheme Test SwitchTheme with default 1`] = `
.c4 {
  cursor: default;
  border: 0;
  background-color: #000;
  color: #fff;
  font-size: 16px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1:after {
  left: 0;
  right: 0;
  bottom: 0;
  content: '';
  position: absolute;
  -webkit-transition: -webkit-transform 200ms cubic-bezier(0,0,0.2,1) 0ms;
  -webkit-transition: transform 200ms cubic-bezier(0,0,0.2,1) 0ms;
  transition: transform 200ms cubic-bezier(0,0,0.2,1) 0ms;
  border-bottom: 2px solid #90caf9;
  pointer-events: none;
}

.c2 {
  -webkit-user-modify: read-only !important;
  padding: 0;
  margin: 0;
  border: 0;
  cursor: default;
}

.c2::-webkit-inner-spin-button,
.c2::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c2::-webkit-inner-spin-button,
.c2::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c2:focus {
  background-color: rgb(153,200,255);
  color: rgb(0,0,0);
  outline: none;
}

.c3 {
  cursor: default;
}

<div>
  <div
    class=""
  >
    <label
      class=""
      for="date"
    >
      Datetime picker:
    </label>
    <div
      class="c0"
    >
      <div
        class="c1"
      >
        <input
          class="c2"
          type="text"
          value="14"
        />
        <div
          class="c3"
        >
          /
        </div>
        <input
          class="c2"
          type="text"
          value="12"
        />
        <div
          class="c3"
        >
          /
        </div>
        <input
          class="c2"
          type="text"
          value="2020"
        />
      </div>
      <button
        class="c4"
        data-testid="app-button"
      >
        Calendar icon
      </button>
    </div>
  </div>
</div>
`;
